### ISSUE TITLE
Add an enrolment event into withExperiment

### ISSUE THREAD

**Issue by bobsilverberg** [mozilla/addons-frontend#8363]
I met with @felixlawrence today to discuss the A/B test for #8137 and he advised that we should be sending an allocation event to GA whenever a user is allocated into a cohort. This is not something that I am currently doing in #8348, where I am only sending an event when a user views an install button for a non-recommended add-on.

As this is something that we should have for any A/B test, I think it makes sense to add it into the `withExperiment` HOC so that an event is sent whenever a user is allocated to a cohort (which signifies them being enrolled into the experiment).

I plan to add this and I will make that change a blocker for #8348.

@willdurand and/or @kumar303, do you have any thoughts or opinions about this?

**Comment by willdurand** [comment]
Sorry, I am curious: I'd like to know more about this allocation metric. This reminds me that we already conducted an experiment on the install button: https://github.com/mozilla/addons-frontend/pull/6675.

It looks similar. The setup was to always send the variant (it was the color of the button) when the button is displayd. We track the install/click events already, so based on the variant, one can deduce the number of installs in GA.

**Comment by felixlawrence** [comment]
Whenever someone gets enrolled into an experiment and allocated to a branch, we need to record this so that when analysing the data we can tell how many people were allocated to each branch. If this fact is not automatically recorded immediately, then it's fairly common that experiment-breaking bugs creep in. These bugs usually take the form of us detecting fewer enrollments in one branch because that branch churned users before they sent the enrollment events. By sending the enrollment events automatically and immediately, it reduces the risk of this happening.

**Comment by kumar303** [comment]
> I met with @felixlawrence today to discuss the A/B test for #8137 and he advised that we should be sending an allocation event to GA whenever a user is allocated into a cohort. This is not something that I am currently doing in #8348

Doesn't this call to [`setDimension()`](https://github.com/mozilla/addons-frontend/pull/8348/files#diff-4853f534279b024e2f35ee066e68850dR115) give you the data you need to infer allocation?

> As this is something that we should have for any A/B test, I think it makes sense to add it into the `withExperiment` HOC so that an event is sent whenever a user is allocated to a cohort (which signifies them being enrolled into the experiment).

I am not totally clear on what you mean here. Do you mean that `withExperiment` should automatically set the dimension to the variant (A or B)?

It might be easier if you used #8348 to implement what you're thinking of. Once that has gone through review and is working, you could follow-up to extract it into a general purpose `withExperiment` feature.

I get why we delete old experiments but it's pretty hard to dig up past usage of `withExperiment()`. I'm not sure how we have addressed this in the past.

**Comment by bobsilverberg** [comment]
> > I met with @felixlawrence today to discuss the A/B test for #8137 and he advised that we should be sending an allocation event to GA whenever a user is allocated into a cohort. This is not something that I am currently doing in #8348
> 
> Doesn't this call to [`setDimension()`](https://github.com/mozilla/addons-frontend/pull/8348/files#diff-4853f534279b024e2f35ee066e68850dR115) give you the data you need to infer allocation?
> 

I believe the issue is that simply setting the dimension does not trigger an event, so, theoretically, a user could be enrolled in a cohort but not have any events recorded, and I think that would go against @felixlawrence's recommendation that we always have an enroll event. @felixlawrence Can you confirm that this is a concern of yours?

> > As this is something that we should have for any A/B test, I think it makes sense to add it into the `withExperiment` HOC so that an event is sent whenever a user is allocated to a cohort (which signifies them being enrolled into the experiment).
> 
> I am not totally clear on what you mean here. Do you mean that `withExperiment` should automatically set the dimension to the variant (A or B)?
> 

No, what I mean is that `withExperiment` should send a GA tracking event whenever a user is enrolled in a cohort (i.e., when the cookie is first set for a user). 

> It might be easier if you used #8348 to implement what you're thinking of. Once that has gone through review and is working, you could follow-up to extract it into a general purpose `withExperiment` feature.
> 

I think the code I'm talking about needs to live in `withExperiment` because that's where the enrollment occurs. If we want to be sure to send an event the moment a user is enrolled, then that code needs to be in `withExperiment`. I.e., I cannot add it into `InstallWarning` in PR #8348.

That's the reason I opened this issue. While I could make those changes to `withExperiment` in #8348 rather than in a separate PR for this issue, that would make the diff for #8348 larger, so it seems a better idea to deal with this concern separately in a PR for this issue. Do you disagree @kumar303?
 
> I get why we delete old experiments but it's pretty hard to dig up past usage of `withExperiment()`. I'm not sure how we have addressed this in the past.

It takes some code archaeology to find examples of where we've used `withExperiment` in the past, but I did find a few examples and my belief is that we *have not* addressed this specific concern in the past.

**Comment by kumar303** [comment]
Ah, ok, I think I understand. We need an enrollment event to analyze how many people see the [add-on warning](https://github.com/mozilla/addons-frontend/issues/8347) and *do not* click the install button, right? In that case, it makes sense to add the enrollment event to `withExperiment` first in a separate PR.

**Comment by felixlawrence** [comment]
> We need an enrollment event to analyze how many people see the add-on warning and do not click the install button, right?
Yes, absolutely - and getting this number correct (for both branches) is pretty much just as important as accurately tracking the number who _do_ click the button.

**Comment by bobsilverberg** [comment]
Marking this as qa:not needed because there isn't really any way to test this without an existing experiment. It can be tested once the A/B test for the install warning lands, which is coming in #8137. I'll add some testing info about this to that issue.
